disk: add WriteAt

yichengq opened the pull request.

yichengq committed:
disk: add WriteAt

xiang90 commented on disk/disk.go:
```diff
@@ -1 +1,97 @@
+// It returns the number of bytes written from p (0 <= n <= len(p)) and
+// any error encountered that caused the write to stop early.
+// WriteAt must return a non-nil error if it returns n < len(p).
+// TODO(xiang90): []byte or io.reader?
```
remove this todo please

xiang90 commented on disk/disk.go:
```diff
@@ -1 +1,97 @@
+)
+
+// WriteAt writes len(p) bytes from p to the file at path at offset off.
+// It returns the number of bytes written from p (0 <= n <= len(p)) and
```
can we simply copy what file.WriteAt says?

xiang90 commented on disk/disk.go:
```diff
@@ -1 +1,97 @@
+		poff := i - pidx*psize
+		wbuf := make([]byte, 0)
+
+		// read out the head of data in block
```
well... i would 
1. find the index of the first block to write
2. find the index of the last block to write
3. if both first,last is in flen -> a
4. if first is in flen, but last is not -> b
5. if first is not in flen -> c

and  i would prefer to handle first,last out of for loop.

yichengq committed:
improve based on comments

yichengq commented:
address the comments.

@xiang90 I tried the solution you mentioned, and i found it does make logic clear, and it needs helper func to write(it needs to handle first/last block special case). Later I found that 1. represent the data that needs to be written 2. write data while handle first, last out of for loop seems more clear.

yichengq committed:
remove buffer struct

Use basic slice instead of buffer struct. This makes code simpler
and avoids pre-optimization.

yichengq committed:
refine var names

yichengq committed:
add helper func fillBlock

yichengq commented:
@xiang90 use fillBlock func to remove duplicate code and make it simpler.

@xiang90 @yunxing @yutongp 
If WriteAt writes 10MB data starting from 10MB offset at empty file, and it fails after writing 5MB all-0 block into the file, what `n` should it return? @yunxing How do you make this case in local file system last time?

xiang90 commented on disk/disk.go:
```diff
@@ -1 +1,98 @@
+	// payload size
+	psize := bsize - crc32Len
+
+	f, err := os.OpenFile(path, os.O_WRONLY, 0600)
```
you still need to close the f.

xiang90 commented on disk/disk.go:
```diff
@@ -1 +1,98 @@
+
+	n := 0
+	stIdx, endIdx := st/psize, end/psize
+	for i := stIdx; i < endIdx; i++ {
```
well... i still think no branch in for loop would be much clearer.

yunxing commented:
@yichengq If it failed at writing 0 bytes, 0 should be returned since no meaningful user data has been written yet.

yichengq committed:
close file when WriteAt finishes

yichengq committed:
handle special case outside of for loop

yichengq committed:
return correct byte number written

yichengq commented:
all comments are addressed.

xiang90 commented on disk/disk.go:
```diff
@@ -1 +1,121 @@
+	return fi.Size()
+}
+
+func max(a, b int) int {
```
can we move this util.go?

xiang90 commented on disk/disk.go:
```diff
@@ -1 +1,121 @@
+}
+
+// blockSize returns the block size of the file at given path.
+func blockSize(path string) int64 {
```
move this one to util.go? blockSize -> getBlockSize

xiang90 commented on disk/disk.go:
```diff
@@ -1 +1,121 @@
+	return 4096
+}
+
+func getFileLength(f *os.File) int64 {
```
move this one to util.go?
